### ISSUE TITLE
found a by-ref issue when analysing legacy code

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#1413]
# issues
1) by-ref variable declared as int before closure that has no operations other than int operators on an array generated by `range(1,10);` erroneously reports `MixedArrayOffset` [regardless of typehint](https://psalm.dev/r/c8feab61fe)
2) `++$i` w/ `MixedOperand` on the by-ref [likely as virtue of being by-ref](https://psalm.dev/r/5aa14c55e1)
3) wording of `UnusedVariable` is inaccurate for treatment of [by-ref](https://psalm.dev/r/237ed9e97f) variable that is used earlier in the closure, and inaccurate for treatment of [no-ref](https://psalm.dev/r/5aa14c55e1)

# repro
* https://3v4l.org/hRRC8
```php
<?php
$a = range(1, 10);
$i = 0;

$b = array_map(
  function (int $c) use (&$i, $a) : string {
    $d = (string) $c . ':' . (string) $a[$i];
    
    $first = 0 === $i;
    
    ++$i;
    
    return $first ? 'first' : $d;
  },
  $a
);

foreach ($b as $e) {
  echo $e, "\n";
}
```
## Actual
> Psalm output (using commit 564da61): 
> 
> INFO: MixedArrayOffset - 7:39 - Cannot access value on variable $a using mixed offset
> 
> INFO: MixedOperand - 11:7 - Left operand cannot be mixed
> 
> INFO: UnusedVariable - 11:7 - Variable $i is never referenced

## Expected
1) If the offset is from a by-ref variable in a closure, perhaps change the message to suggest the variable be [type-hinted](https://psalm.dev/r/372010bffc) or [type-cast](https://psalm.dev/r/242c9d1476) first ?
2) `MixedOperand` in this example appears to be a direct result of it being by-ref, and is resolved either by not passing it as by-ref or by type-hinting/casting as above.
3) Either reword an `UnusedVariable` to something to the effect of `Variable $i is not referenced after this point` as it is referenced earlier in the closure, or specifically suppressing the issue if `$i` is by-ref ***and*** referenced earlier in the closure.

**Comment by bugreportuser** [comment]
Related: https://github.com/vimeo/psalm/issues/678#issuecomment-463710808

**Comment by muglug** [comment]
The only thing that's wrong here is the `UnusedVariable` - otherwise Psalm is working as expected - that is, it always treats by-ref params in closure `use`  as `mixed`

**Comment by SignpostMarv** [comment]
@muglug does psalm have the stuffs for conditionally altering the error message for user-friendliness ?

**Comment by muglug** [comment]
Nope
